openapi-core/src/main/java/io/ballerina/openapi/core/generators/client/AbstractFunctionSignatureGenerator.java

```diff
@@ -0,0 +1,285 @@
+import io.swagger.v3.oas.models.OpenAPI;
+import io.swagger.v3.oas.models.Operation;
+import io.swagger.v3.oas.models.parameters.Parameter;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static io.ballerina.compiler.syntax.tree.AbstractNodeFactory.createSeparatedNodeList;
+import static io.ballerina.compiler.syntax.tree.AbstractNodeFactory.createToken;
+import static io.ballerina.compiler.syntax.tree.SyntaxKind.CLOSE_PAREN_TOKEN;
+import static io.ballerina.compiler.syntax.tree.SyntaxKind.COMMA_TOKEN;
+import static io.ballerina.compiler.syntax.tree.SyntaxKind.OPEN_PAREN_TOKEN;
+import static io.ballerina.openapi.core.generators.client.diagnostic.DiagnosticMessages.OAS_CLIENT_100;
+import static io.ballerina.openapi.core.generators.common.GeneratorConstants.HEADERS;
+import static io.ballerina.openapi.core.generators.common.GeneratorUtils.extractReferenceType;
+
+/**
+ * This class represents the abstract function signature generator which is used to implement
+ * the common functionalities of the function signature generators.
+ *
+ * @since 2.1.1
+ */
+public abstract class AbstractFunctionSignatureGenerator implements FunctionSignatureGenerator {
+    OpenAPI openAPI;
+    Operation operation;
+    List<ClientDiagnostic> diagnostics  = new ArrayList<>();
+    FunctionReturnTypeGeneratorImp functionReturnTypeGenerator;
+    private final String httpMethod;
+    private final String path;
+    private String headersParamName = HEADERS;
+
+    private boolean hasDefaultHeader = false;
+    private boolean hasHeadersParam = false;
+    private boolean hasQueriesParam = false;
+
+    protected AbstractFunctionSignatureGenerator(Operation operation, OpenAPI openAPI, String httpMethod,
+                                                 String path) {
+        this.operation = operation;
+        this.openAPI = openAPI;
+        this.httpMethod = httpMethod;
+        this.path = path;
+        this.functionReturnTypeGenerator = new FunctionReturnTypeGeneratorImp(operation, openAPI, httpMethod);
+    }
+
+    @Override
+    public Optional<FunctionSignatureNode> generateFunctionSignature() {
+        List<Parameter> parameters = operation.getParameters();
+        ParametersInfo parametersInfo = getParametersInfo(parameters);
+
+        if (Objects.isNull(parametersInfo)) {
+            return Optional.empty();
+        }
+
+        List<Node> paramList = getParameterNodes(parametersInfo);
+        SeparatedNodeList<ParameterNode> parameterNodes = createSeparatedNodeList(paramList);
+
+        // 3. return statements
+        FunctionReturnTypeGeneratorImp functionReturnType = getFunctionReturnTypeGenerator();
+        Optional<ReturnTypeDescriptorNode> returnType = functionReturnType.getReturnType();
+        diagnostics.addAll(functionReturnType.getDiagnostics());
+        if (returnType.isEmpty()) {
+            return Optional.empty();
+        }
```

Comment:
As per the Optional documentation [1] this is not necessary as if the optional is empty this would anyway return an empty optional.
```suggestion
```

[1] - https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#map-java.util.function.Function-

Reply:
Removed via [a25606b](https://github.com/ballerina-platform/openapi-tools/pull/1768/commits/a25606bf532fa022f92f7bd1256dc6cdeb21d9cf)